### ISSUE TITLE
Issues with JSON payloads

### ISSUE THREAD

**Issue by oshoemaker** [mcollina/ascoltatori#129]
This was introduced in commit: 5f6a14d0134346ecac85247e296058590a17fe10 (aka v1.0.0)

It seems that the offending line is in ascoltatori.js (line 82):

+    result = new module.exports.JSONAscoltatore(result)
+      .once("error", done);

There is no test coverage for this (or any JSON payloads?).  

The error is:

/Users/owen/Repos/frontend-server/node_modules/ascoltatori/lib/ascoltatori.js:87
      result.once("ready", function() {
            ^
TypeError: Cannot read property 'once' of undefined
    at Immediate._onImmediate (/Users/owen/Repos/frontend-server/node_modules/ascoltatori/lib/ascoltatori.js:87:13)
    at processImmediate [as _immediateCallback] (timers.js:367:17)

It appears that you are trying to call a method on a object returned but I am not sure that this will work.  I would write a test case and commit a fix but unfortunately I don't have time at the moment.  If somebody else could help with a fix that would be appreciated.

**Comment by mcollina** [comment]
Yes, `build` has no test.

The problem is here: https://github.com/mcollina/ascoltatori/blob/master/lib/decorator_ascoltatore.js#L58-L64. I think the fix is straightforward, would you mind sending a PR?

**Comment by oshoemaker** [comment]
Sorry, the fix does not seem straight forward to me.  Can you add any background here?

**Comment by mcollina** [comment]
on and once should return this
Il giorno ven 21 ago 2015 alle 23:50 oshoemaker <notifications@github.com>
ha scritto:

**Comment by oshoemaker** [comment]
Sorry, I am not as familiar with the project as you.  As far as I can tell there is no on or once function.  It would seem that the ascoltatore event emitter needs to be wired up to the object as it doesn’t exist in the ascoltatore object that is being used.  I honestly don’t have time to dig through this right now.  I might be able to get to it in a couple months but I can simply work around it.

**Comment by mcollina** [comment]
It's a two line fix. You  a 'return this' entry in the code I linked above at the end of each highlighted function.

**Comment by oshoemaker** [comment]
As far as I can tell this will not fix the issue either.  Tested and still throwing errors:

DecoratorAscoltatore.prototype.on = function(event, callback) {
  this._ascoltatore.on(event, callback);
	return this;
};

DecoratorAscoltatore.prototype.once = function(event, callback) {
  this._ascoltatore.once(event, callback);
	return this;
};

**Comment by oshoemaker** [comment]
I attempted to add the fix but ended up with the object being undefined.  Any suggestions?  Here is the output:

TypeError: Cannot read property 'unsubscribe' of undefined
    at /Users/owen/Repos/socket-server/lib/ShardedAscoltatori.js:151:14
    at /Users/owen/Repos/socket-server/node_modules/async/lib/async.js:122:13
    at _each (/Users/owen/Repos/socket-server/node_modules/async/lib/async.js:46:13)
    at Object.async.each (/Users/owen/Repos/socket-server/node_modules/async/lib/async.js:121:9)
    at ShardedAscoltatori.unsubscribe (/Users/owen/Repos/socket-server/lib/ShardedAscoltatori.js:144:8)
    at SocketServer.stop (/Users/owen/Repos/socket-server/app.js:67:20)
    at process.<anonymous> (/Users/owen/Repos/socket-server/app.js:157:18)
    at emitOne (events.js:77:13)
    at process.emit (events.js:169:7)
    at process.exit (node.js:737:17)

Get a similar issues for subscribes as well:

UncaughtException: TypeError: Cannot read property 'subscribe' of undefined
    at /Users/owen/Repos/socket-server/lib/ShardedAscoltatori.js:125:14
    at /Users/owen/Repos/socket-server/node_modules/async/lib/async.js:122:13
    at _each (/Users/owen/Repos/socket-server/node_modules/async/lib/async.js:46:13)
    at Object.async.each (/Users/owen/Repos/socket-server/node_modules/async/lib/async.js:121:9)
    at ShardedAscoltatori.subscribe (/Users/owen/Repos/socket-server/lib/ShardedAscoltatori.js:117:8)
    at /Users/owen/Repos/socket-server/app.js:101:24
    at fn (/Users/owen/Repos/socket-server/node_modules/async/lib/async.js:638:34)
    at Immediate._onImmediate (/Users/owen/Repos/socket-server/node_modules/async/lib/async.js:554:34)
    at processImmediate [as _immediateCallback] (timers.js:371:17) +9ms
on exit socketServer=[object Object]

**Comment by mcollina** [comment]
What is `ShardedAscoltatori` this is definitely not in here.

**Comment by goliatone** [comment]
I can confirm this is still an issue with **ascoltatori@^1.1.0**. 

A working codebase using **0.20.2** upgraded to the latest version stopped working, generating the following error:
```
TypeError: Cannot read property 'once' of undefined
    at Immediate._onImmediate (/Users/goliatone/Dev/dashboard/node_modules/ascoltatori/lib/ascoltatori.js:87:13)
    at processImmediate [as _immediateCallback] (timers.js:358:17)
```

The suggested fix prevented the error but the connection timesout from time to time. Using the **0.20.2** version timeouts does not seem to happen.

**Comment by mcollina** [comment]
Fixed in v2.0.0
